### PR TITLE
Implement arbitrary penalties for SNLLS & Pathway amplitudes contstraint

### LINKED ISSUE #76: Scale invariability of pathway amplitudes and redundancy with V0
In the design of DeerLab seems to be some redundancy and invariability issues related to how the dipolar pathway amplitudes are defined and handled. 

### Scale invariability of dipolar pathways
In experiment models (except ``ex_4pdeer``) where the unmodulated contribution is parametrized (``Lam0``), the absolute values of the pathway amplitudes result in a scale-invariant response in DeerLab. For example the simulations for `V1` and `V2` here

        V1= dl.dipolarkernel(t,r,dl.ex_5pdeer([0.8, 0.4, 0.2, 4]))@P
        V2= dl.dipolarkernel(t,r,dl.ex_5pdeer([0.4, 0.2, 0.1, 4]))@P
        np.array_equal(V1,V2)
        >> True

will result in exactly the same signal. The reason behind this is the default behavior of `dipolarkernel`, which normalizes the dipolar kernel after calculations (controlled by the option ``renormalize``). BTW there is an additional option called ``renormpaths`` which has no effect on the output and will be deprecated in a future PR). 

As expected by setting `renormalize=False` the scale-invariability of the dipolar pathways is lifted

        V1= dl.dipolarkernel(t,r,dl.ex_5pdeer([0.8, 0.4, 0.2, 4]),renormalize=False)@P
        V2= dl.dipolarkernel(t,r,dl.ex_5pdeer([0.4, 0.2, 0.1, 4]),renormalize=False)@P
        np.array_equal(V1,V2)
        >> False
  
### Redundancy with V0

Setting `renormalize=False` reveals a second issue: the absolute values of the pathway amplitudes and the dipolar signals scale 'V0' become redundant. For example the simulations for `V1` and `V2` here

        V0 = 1
        V1= V0*dl.dipolarkernel(t,r,dl.ex_5pdeer([0.8, 0.4, 0.2, 4]),renormalize=False)@P
        V0 = 2
        V2= V0*dl.dipolarkernel(t,r,dl.ex_5pdeer([0.4, 0.2, 0.1, 4]),renormalize=False)@P
        np.array_equal(V1,V2)
        >> True

the signals are again equal. Only by reinstating the kernel normalization can these be distinguished again. 

### Summary
- Normalization of the dipolar kernel (i.e. enforcing `K(0,r)=1`) is required for `V0` in the separation `V(t)=V0*Vintra(t)*Vinter(t)` to be distinguishable.
- Under normalization, pathway amplitudes in models of the form `V0*(Lam0 + sum_i lam_i*Vintra_i)` exhibit scale invariability.

The question now is: what is the optimal and correct way to treat this? 
On one hand, pathway amplitudes are probabilistic quantities so it would make sense to think of them as being confined to a range [0,1]. On the other hand, we could imagine them relative contributions to `V0`, i.e. such that the sum over all dipolar pathways equals `V0`. This second interpretation might be more efficient due to the fact that the ``ex_4pdeer`` does not suffer from the issues described here, as the contributions are also define in a relative way.  

This requires further discussion.

Here is a full working example with a graphical representation of these issues:

````python
import deerlab as dl
import numpy as np 
import matplotlib.pyplot as plt 

t = np.linspace(-0.1,7,200)
r = np.linspace(2,6,200)
P = dl.dd_gauss(r,[3,0.2])

def V5pdeer_default(lam):
    K = dl.dipolarkernel(t,r,dl.ex_5pdeer([lam[0],lam[1],lam[2],4.23]),lambda lam:dl.bg_hom3d(t,50,lam),renormpaths=False)
    return K@P

def V5pdeer(lam):
    K = dl.dipolarkernel(t,r,dl.ex_5pdeer([lam[0],lam[1],lam[2],4.23]),lambda lam:dl.bg_hom3d(t,50,lam),renormalize=False,renormpaths=False)
    return K@P

plt.figure(figsize=[7,6])
plt.subplot(221)
plt.plot(t,V5pdeer_default([0.8,0.6,0.4]),'-',linewidth=2)
plt.plot(t,V5pdeer_default([0.4,0.3,0.2]),'--',linewidth=2)
plt.xlabel('t [us]')
plt.ylabel('V(t)')
plt.title('$V_0$=1, Kernel normalized')
plt.legend(['[$\Lambda_0$,$\lambda_1$,$\lambda_2$]','$\\frac{1}{2}$[$\Lambda_0$,$\lambda_1$,$\lambda_2$]'])

plt.subplot(222)
plt.plot(t,V5pdeer([0.8,0.6,0.4]),'-',linewidth=2)
plt.plot(t,V5pdeer([0.4,0.3,0.2]),'--',linewidth=2)
plt.xlabel('t [us]')
plt.ylabel('V(t)')
plt.title('$V_0$=1, Kernel not normalized')


plt.subplot(223)
V0 = 2
plt.plot(t,V5pdeer_default([0.8,0.6,0.4]),'-',linewidth=2)
plt.plot(t,V0*V5pdeer_default([0.4,0.3,0.2]),'--',linewidth=2)
plt.xlabel('t [us]')
plt.ylabel('V(t)')
plt.title('$V_0$=2, Kernel normalized')

plt.subplot(224)
V0 = 2
plt.plot(t,V5pdeer([0.8,0.6,0.4]),'-',linewidth=2)
plt.plot(t,V0*V5pdeer([0.4,0.3,0.2]),'--',linewidth=2)
plt.xlabel('t [us]')
plt.ylabel('V(t)')
plt.title('$V_0$=2, Kernel not normalized')

plt.tight_layout()
plt.show()
````
issue

### LINKED ISSUE #103: snnls: double addition of regularization term
As noted by @mtessmer, `ResidualFcn` in `snnls.py` seems to be adding the regularization term twice, once prior to calling `_augment`, and then again in `_augment` itself. It should only be once.

Here is the place in ``snnls.py``:
```python
if includePenalty:
    penalty = alpha*L@linfit
    # Augmented residual
    res = np.concatenate((res, penalty))     #    <---  first addition
    res, _ = _augment(res, [], regtype, alpha, L, linfit, huberparam, Nnonlin)    #   <---- second addition
```

### PR DESCRIPTION
**Pull request by luisfabib**
## Description

This is a major upgrade to `snlls.py` and `fitmodel.py`. It improves `snlls.py` to be able to handle arbitrary penalties defined by outer functions or scripts. This allows us to implement a penalty in `fitmodel.py` that constrains the pathway amplitudes and makes the separation of the `Lam0` and `V0` parameters identifiable, strongly improving the fit accuracy and suppressing most uncertainty in fits of multi-pathway models such as 5-pulse DEER, etc.

## Changelog:

`snlls.py`
- Implement an option to pass arbitrary penalty function residuals to be included in the SNLLS algorithm. 
- Remove double smoothness penalization in the residual (#103).
- Return the uncertainty quantification of linear and nonlinear parts as separate objects.
- Improve uncertainty analysis by including correlations between linear and non-linear parameters.
- Major code refactoring.
- Improved behavior of scale determination.

`fitmodel.py`
- Implement constraint `Lam0+sum(lam)<=1` to ensure identifiability of `Lam0` and `V0` during SNLLS optimization (#76). 
- Refactor the uncertainty handling code to adapt to the new changes in `snlls.py`.

## Related Issues:

Closes #103
Closes #76

### CODE REVIEW

**Comment by luisfabib** [comment]
The new keyword argument in `snlls.py` to pass custom penalties is called `custom_penalty`. We might want to change this, as it was the quick name I chose during development. 

Another notation issue might be the difference between the built-in smoothness penalty (needed to ensure the linear problem is correctly computed) and any custom penalty passed by the user. A way to refer to them would be linear-penalty (for the smoothness penalty applied to the linear problem) and non-linear penalty (the user-defined penalty applied to the non-linear problem).  

We need to adjust this before merging.

File: deerlab/snlls.py [hunk 1]
```diff
@@ -247,9 +247,9 @@ def snlls(y, Amodel, par0, lb=None, ub=None, lbl=None, ubl=None, nnlsSolver='cvx
     nonNegativeOnly = (np.all(lbl == 0)) and (np.all(np.isinf(ubl)))
 
 
+    # Use an arbitrary axis
+    ax = np.arange(1, Nlin+1)
     if includePenalty:
```
Comment (stestoll) [inline deerlab/snlls.py:252]: Why do you set an L even if includePenalty is False?

Comment (stestoll) [inline deerlab/snlls.py:252]: `includePenalty` is now ambiguous - what about calling it `includeRegularization`

File: deerlab/snlls.py [hunk 2]
```diff
@@ -284,6 +284,31 @@ def snlls(y, Amodel, par0, lb=None, ub=None, lbl=None, ubl=None, nnlsSolver='cvx
...
+        """
+        Linear problem
+        ------------------
+        Solves the linear subproblem of the SNLLS objective function via linear LSQ 
+        constrained, unconstrained, or regularized.
+        """
+        global alpha
+        
+        # Regularization components
+        if optimize_alpha:
```
Comment (stestoll) [inline deerlab/snlls.py:299]: The calls to `selregparam` and `lsqcomponents` (a few lines down) should only happen if regularization is requested.

File: deerlab/snlls.py [hunk 3]
```diff
@@ -422,17 +456,38 @@ def ci(coverage,ptype='full'):
...
+    uq_subset.nparam = len(subset)
+
+    # Get requested confidence interval of joined parameter set
+    uq_subset.ci = lambda coverage: uq_full.ci(coverage)[subset, :]
+    uq_subset.percentile = lambda p: uq_full.percentile(p)[subset]
+
+    return uq_subset
+#===========================================================================
+
+def reg_penalty(regtype, alpha, L, x, eta, Nnonlin):
```
Comment (stestoll) [inline deerlab/snlls.py:490]: Rename to `augment_residual` or similar.

File: deerlab/snlls.py [hunk 4]
```diff
@@ -363,57 +402,52 @@ def ResidualsFcn(p):
...
+        
+        # Jacobian (non-linear part)
+        Jnonlin = Jacobian(ResidualsFcn,nonlinfit,lb,ub)
 
-        Jlin = Afit
-        J = np.concatenate((Jnonlin, Jlin),1)
+        # Jacobian (linear part)
+        Jlin = np.zeros((len(res),len(linfit)))
+        Jlin[:len(y),:] = Amodel(nonlinfit)
+        Jlin[len(res)-Nlin:,:] = reg_penalty(regtype, alpha, L, linfit, huberparam, Nnonlin)[1]
```
Comment (stestoll) [inline deerlab/snlls.py:423]: Does this copy the augmented residual into `Jlin`? `reg_penalty` returns the residual first and the Jacobian second, right?

**Review by stestoll — CHANGES_REQUESTED** [review]

File: deerlab/snlls.py (hunk 2 above)
Comment (luisfabib) [reply deerlab/snlls.py:299]: The call to `selregparam` happens only when regularization is requested (in which case `optimize_alpha=False`).  The logic statements are defined in `ResidualsFcn` in lines 328-351. 

The call to `lsqcomponents ` is required even when regularization is not requested. That function handles other things such as global optimization and proper preparation of the arguments of NNLS.  In the case that regularization is not needed, `alpha=0` and `L=eye(N.N)` so that the penalty is not included even though they are passed to `lsqcomponents `.

File: deerlab/snlls.py (hunk 1 above)
Comment (luisfabib) [reply deerlab/snlls.py:252]: An arbitrary `L` is defined when `includePenalty=False` to avoid a reference-error later in line 303
```python
AtA, Aty = dl.lsqcomponents(y, A, L, alpha, weights, regtype=regtype)
````

File: deerlab/snlls.py (hunk 4 above)
Comment (luisfabib) [reply deerlab/snlls.py:423]: Note the `[1]` at the end of the call. It is copying the second output argument, i.e. the Jacobian.

File: deerlab/snlls.py (hunk 3 above)
Comment (luisfabib) [reply deerlab/snlls.py:490]: `augment_residual ` is no longer an appropriate name as the function no longer augments the residual and Jacobian. Instead it is returning the residual and Jacobian subsets corresponding to the regularization penalty. The augmentation or copying is done later in the main code.

File: deerlab/snlls.py (hunk 1 above)
Comment (luisfabib) [reply deerlab/snlls.py:252]: I agree that `includePenalty ` is ambiguoug, however `includeRegularization` is a bit too wordy for a keyword argument.

Maybe something like `regpenalty`?

File: deerlab/snlls.py [hunk 5]
```diff
@@ -16,7 +16,7 @@
 from deerlab.classes import UncertQuant, FitResult
 
 def snlls(y, Amodel, par0, lb=None, ub=None, lbl=None, ubl=None, nnlsSolver='cvx', reg='auto', weights=1,
-          regtype='tikhonov', regparam='aic', multistart=1, regorder=2, alphareopt=1e-3,
+          regtype='tikhonov', regparam='aic', multistart=1, regorder=2, alphareopt=1e-3, custom_penalty=None,
```
[the remark this answers is not in the document]
Comment (luisfabib) [reply deerlab/snlls.py]: Done

File: deerlab/snlls.py [hunk 6]
```diff
@@ -293,54 +318,68 @@ def ResidualsFcn(p):
...
+        regparam_prev = alpha
 
-        # Solve the linear least-squares problem
-        result = linSolver(AtA, Aty)
-        linfit = parseResult(result)
-        linfit = np.atleast_1d(linfit)
         # Evaluate full model residual
         yfit = A@linfit
+
+        scales = []
```
[the remark this answers is not in the document]
Comment (luisfabib) [reply deerlab/snlls.py]: Done

File: deerlab/snlls.py (hunk 1 above)
Comment (stestoll) [reply deerlab/snlls.py:252]: That's not clean. I think in the absence of a reg term, `lsqcomponents` should only be called wity `y` and `A`.

Comment (stestoll) [reply deerlab/snlls.py:252]: Fine

File: deerlab/snlls.py (hunk 2 above)
Comment (stestoll) [reply deerlab/snlls.py:299]: The `alpha=0`and `L=eye(N,N)` then better goes into the defaults of the argument list of `lsqcomponents`.

File: deerlab/snlls.py (hunk 1 above)
Comment (luisfabib) [reply deerlab/snlls.py:252]: My bad, I thought `includePenalty` was a keyword argument, but it is just a variable. I renamed it to `includeRegularization` as you proposed.

File: deerlab/snlls.py [hunk 7]
```diff
@@ -293,54 +318,68 @@ def ResidualsFcn(p):
         non-linear least-squares solver. 
         """
 
-        nonlocal par_prev, check, regparam_prev, linfit, alpha
+        nonlocal par_prev, check, regparam_prev, scales, linfit
+        global alpha
```
[the remark this answers is not in the document]
Comment (luisfabib) [reply deerlab/snlls.py]: Yes, I refactored the code to do this.

File: deerlab/snlls.py (hunk 2 above)
Comment (luisfabib) [reply deerlab/snlls.py:298]: I did so. `lsqcomponents` now sets the `alpha` to zero and `L` to a diagonal matrix if not specified.

File: deerlab/snlls.py (hunk 1 above)
Comment (luisfabib) [reply deerlab/snlls.py]: See next comment.

File: deerlab/lsqcomponents.py [hunk 8]
```diff
@@ -1,11 +1,14 @@
...
 
-def lsqcomponents(V,K,L,alpha,weights, regtype = 'tikh',huberparam = 1.35):
+def lsqcomponents(V,K,L=None,alpha=0,weights=1, regtype = 'tikh',huberparam = 1.35):
 # ==============================================================================================
 
     # Prepare
     nr = np.shape(K)[1]
 
+    if L is None:
+        L = np.eye(nr)
```
Comment (stestoll) [inline deerlab/lsqcomponents.py:10]: Maybe set L to zero?

Comment (luisfabib) [reply deerlab/lsqcomponents.py]: If the `L=None` then the `lsqcomponents` function now returns directly the `AtA` and `AtV` components without the regularization terms.

**Review by stestoll — APPROVED** [review]
